spg/Animal/lib/RaceHorse.pm

```diff
@@ -0,0 +1,197 @@
+package RaceHorse;
+
+use 5.006;
+use strict;
+use warnings;
+
+=head1 NAME
+
+RaceHorse - How many races can it win before going to the glue factory?
+
+=head1 VERSION
+
+Version 0.01
+
+=cut
+
+our $VERSION = '0.01';
+
+use parent qw(Horse);
+use JSON qw( encode_json decode_json );
+use File::Slurp;
+
+=head1 SYNOPSIS
+
+Quick summary of what the module does.
+
+Perhaps a little code snippet.
+
+    use Horse;
+
+    my $foo = Horse->new();
+    ...
+
+=head1 EXPORT
+
+A list of functions that can be exported.  You can delete this section
+if you don't export anything, such as for a purely object-oriented module.
+
+=head1 SUBROUTINES/METHODS
+
+=head2 get_type
+
+=cut
+
+sub get_type {
+  return "RaceHorse";
+}
+
+=head2 get_voice
+
+=cut
+
+sub get_voice {
+  return "strong neigh";
+}
+
+my $file = "stats.json";
+my @stats = qw( wins places shows losses );
+
+sub new {
+  my $self = shift->SUPER::new(@_);
+
+  my $stats_data;
+
+  if ( -e $file ) {
+    eval {
+      $stats_data = decode_json( scalar( read_file( $file ) ) );
+    };
+    if ( $@ ) {
+      print "Invalid JSON: $@";
+      $stats_data = map { $_ => 0 } @stats;
+    }
+  }
+  else {
+    $stats_data = map { $_ => 0 } @stats;
+  }
+
+  @{$self}{@stats} = @{$stats_data}{@stats};
+
+  $self;
+}
+
+sub won { $_[0]->{wins}++; }
+
+sub placed { $_[0]->{places}++; }
+
+sub showed { $_[0]->{shows}++; }
+
+sub lost { $_[0]->{losses}++; }
+
+sub standings {
+  my ( $self ) = @_;
+  join ', ', map "$self->{$_} $_", @stats;
+}
+
+sub DESTROY {
+
+  my $stats;
+
+  @{$stats}{ keys %{$_[0]} } = @{$_[0]}{ keys %{$_[0]} };
+
+  my $json = encode_json( $stats );

```

Comment:
Right you are, looks like `JSON` restricts this unless you enable it (presumably so that objects can decide how to encode themselves with the `TO_JSON` method).

You could still simply copy the hash like this to get around it:
```perl
my $json = encode_json( { %self } );
```